src/mtcp/mtcp_restart.c

```diff
@@ -850,7 +850,10 @@ unmap_memory_areas_and_restore_vdso(RestoreInfo *rinfo)
       mtcp_abort();
     }
     MTCP_ASSERT(vvar == vvarStart);
-    mtcp_memcpy(vvarStart, rinfo->vvarStart, vvarEnd - vvarStart);
+    // On i386, only the first 4096 bytes are readable. Reading beyond that
+    // results in a bus error.
+    // TODO: Find a way to automatically detect the readable bytes.
+    mtcp_memcpy(vvarStart, rinfo->vvarStart, 4096);//vvarEnd - vvarStart);
```

Comment:
In addition, let's change `4096` to `getpagesize()`.  This should be more future-proof.  The man page says:
> The function getpagesize() returns  ...where "page" is a fixed-length block, the unit for memory allocation and file mapping performed by mmap(2).

Note that we will then need to do:
`#define _DEFAULT_SOURCE`
`#include <unistd.h>`
at the head of the file.  The man page says `_DEFAULT_SOURCE` is valid for glibc-2.19 and beyond, which should handle the recent kernels that cause this bug.

Also, in answer to @rohgarg, I believe this is a kernel bug.  We've seen a similar bug where the kernel removes the write permission for the last page of a memory segment, even though `/proc/*/maps` says that the entire memory segment has read permission.

This also makes me wonder if there is a different fix for this bug.  We could use:
 `mprotect(vvarStart, rinfo->vvarEnd - vvarStart, PROT_READ`
@karya0, could you try out this alternative fix (mprotect instead of modifying mmap)?  If this works, we don't need `getpagesize()`, although I would still keep the additional comments that I'm proposing.